### PR TITLE
fix!: Add DisplacementEvent to fix delta coordinate transformations for drag events

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

This adds `DisplacementEvent` to fix delta coordinate transformations for drag events, to be used instead of `PositionEvent`.
Drag Events now expose the start and end position, as well as the delta, correctly transformed by the camera and zoom.
This also ensures that drag events, once starts, do not get lost if the drag update leaves the component bounds.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [X] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] Yes, this PR is a breaking change.
- [ ] No, this PR is not a breaking change.

### Migration instructions

* if you are using `DragUpdateEvent` events, the `devicePosition`, `canvasPosition`, `localPosition`, and `delta` are deprecated as they are unclear.
* use `xStartPosition` to get the position at the start of the drag event ("from")
* use `xEndPosition` to get the position at the end of the drag event ("to")
* if you want the delta, use `localDelta`. it now already considers the camera zoom. no need to manually account for that
* now you keep receiving drag events for the same component even if the mouse leaves the component (breaking)

## Related Issues

This fixes https://github.com/flame-engine/flame/issues/2843

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: packages/flame/lib/src/events/messages/drag_update_event.dart [hunk 1]
```diff
@@ -1,29 +1,25 @@
...
+          deviceStartPosition: details.globalPosition.toVector2(),
+          deviceEndPosition:
+              details.globalPosition.toVector2() + details.delta.toVector2(),
         );
 
   final int pointerId;
   final Duration timestamp;
-  final Vector2 delta;
-
-  static final _nanPoint = Vector2.all(double.nan);
```
Comment (luanpotter) [inline packages/flame/lib/src/events/messages/drag_update_event.dart:17]: yay!

File: packages/flame/lib/src/components/core/component.dart [hunk 2]
```diff
@@ -700,31 +700,53 @@ class Component {
   /// If your component overrides [renderTree], then it almost certainly needs
   /// to override this method as well, so that this method can find all rendered
   /// components wherever they are.
-  Iterable<Component> componentsAtPoint(
+  Iterable<Component> componentsAtVector2(
```
Comment (spydon) [inline packages/flame/lib/src/components/core/component.dart:703]: Maybe we could keep this as `componentsAtPoint` to reduce the breaking surface?

File: packages/flame/lib/src/components/core/component.dart [hunk 3]
```diff
@@ -700,31 +700,53 @@ class Component {
...
+  ]) {
+    return componentsAtLocation<Vector2>(
+      point,
+      nestedPoints,
+      (transform, point) => transform.parentToLocal(point),
+      (component, point) => component.containsLocalPoint(point),
+    );
+  }
+
+  Iterable<Component> componentsAtLocation<T>(
```
Comment (spydon) [inline packages/flame/lib/src/components/core/component.dart:715]: We need some dartdocs on this one

File: packages/flame/lib/src/events/messages/displacement_event.dart [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+          return null;
+        }
+        return (
+          start: transformedStart,
+          end: transformedEnd,
+        );
+      },
+      // we only trigger the drag start if the component check passes, but
+      // as the user drags the cursor it could end up outside the component
+      // bounds; we don't want the event to be lost, so we bypass the check
```
Comment (spydon) [inline packages/flame/lib/src/events/messages/displacement_event.dart:147]:
```suggestion
      // bounds; we don't want the event to be lost, so we bypass the check.
```

File: packages/flame/lib/src/camera/camera_component.dart [hunk 5]
```diff
@@ -223,22 +221,44 @@ class CameraComponent extends Component {
     return viewport.localToGlobal(viewfinderPosition, output: output);
   }
 
-  final _viewportPoint = Vector2.zero();
```
Comment (spydon) [inline packages/flame/lib/src/camera/camera_component.dart:226]: We will create a lot of new Vector2 if we don't use this one

**Review by spydon — COMMENTED** [review]
Lgtm, just one comment

**Pull request merged**